c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -478,6 +479,10 @@ struct THCCachingAllocator
 
   void free_cached_blocks(int device)
   {
+    // First ensure that all blocks that can't currently be allocated due to
+    // outstanding events are returned to the pool.
+    synchronize_and_free_events(device);
```

Comment:
should this be an option? I am a little worried that some existing app code suddenly much longer delay due to this change, but they may not absolutely need that.

Reply:
This won't slow down existing apps. They already call `cudaFree` in this code path, which waits for all operations on the GPU to finish. The only extra cost is the overhead of the API call, which is pretty small compared to everything else in `free_cached_blocks`.